File: demo/javascript/local_word_count.js
/**
 * Local word count job using the atomic JavaScript API.
 *
 * Prerequisites:
 *   cargo build --release -p atomic-js
 *
 * Run:
 *   node demo/javascript/local_word_count.js
 */

const { Context } = require('../../crates/atomic-js');

const ctx = new Context();

const lines = ctx.parallelize([
  "hello world hello",
  "world foo bar hello",
  "atomic is fast and atomic is small",
  "fast systems need fast tools",
]);

// Classic Spark word count:
//   split lines → emit [word, 1] pairs → aggregate by key
const wordCounts = lines
  .flatMap(line => line.split(/\s+/).map(w => [w, 1]))
  .reduceByKey((a, b) => a + b)
  .collect();

// Sort alphabetically and print
wordCounts.sort((a, b) => a[0].localeCompare(b[0]));

console.log("word                 count");
console.log("----------------------------");
for (const [word, count] of wordCounts) {
  console.log(word.padEnd(20) + " " + count);
}

File: demo/python/local_word_count.py
"""
Local word count job using the atomic Python API.

Prerequisites:
  cd crates/atomic-python && maturin develop --release && cd ../..

Run:
  python demo/python/local_word_count.py
"""

import atomic

ctx = atomic.Context(default_parallelism=4)

# Input: lines of text spread across the dataset
lines = ctx.parallelize([
    "hello world hello",
    "world foo bar hello",
    "atomic is fast and atomic is small",
    "fast systems need fast tools",
])

# Classic Spark word count:
#   split lines → emit (word, 1) pairs → aggregate by key
word_counts = (
    lines
    .flat_map(lambda line: line.split())        # ["hello", "world", "hello", ...]
    .map(lambda w: (w, 1))                      # [("hello", 1), ("world", 1), ...]
    .reduce_by_key(lambda a, b: a + b)          # [("hello", 3), ("world", 2), ...]
    .collect()
)

print(f"{'word':<20} count")
print("-" * 28)
for word, count in sorted(word_counts):
    print(f"{word:<20} {count}")
